readthedocs/vcs_support/backends/git.py

```diff
@@ -68,6 +91,166 @@ def update(self):
+                f"{self.version_identifier}"
+            )
+        # Tags
+        if self.version_type == TAG and self.verbose_name:
+            # A "stable" tag is automatically created with Version.machine=True,
+            # denoting that it's not a tag that really exists.
+            # Because we don't know if it originates from the default branch or some
+            # other tagged release, we will fetch everything.
+            if self.version_machine and self.verbose_name == "stable":
+                if self.version_identifier:
+                    return f"{self.version_identifier}:readthedocs/stable"
+                log.error(
+                    "'stable' version without a commit hash",
+                    version_identifier=self.version_identifier,
+                )
+                return None
+            return f"refs/tags/{self.verbose_name}:refs/tags/{self.verbose_name}"
+
+        if self.version_type == EXTERNAL:
+            # TODO: We should be able to resolve this without looking up in oauth registry
+            git_provider_name = self.project.git_provider_name
+
+            # Remote reference for Git providers where pull request builds are supported
+            if git_provider_name == GITHUB_BRAND:
+                return GITHUB_PR_PULL_PATTERN.format(id=self.verbose_name)
+            if self.project.git_provider_name == GITLAB_BRAND:
+                return GITLAB_MR_PULL_PATTERN.format(id=self.verbose_name)
+
+    def clone_ng(self):
+        # TODO: This seems to be legacy that can be removed.
+        #  If the repository is already cloned, we don't do anything.
+        #  It seems to originate from when a cloned repository was cached on disk,
+        #  and so we can call call .update() several times in the same build.
+        if self.repo_exists():
+            return
+
+        # TODO: This seems to be legacy that can be removed.
+        #  There shouldn't be cases where we are asked to
+        #  clone the repo in a non-clean working directory.
+        #  The prior call to repo_exists() will return if a repo already exist with
+        #  unclear guarantees about whether that even needs to be a fully consistent clone.
+        self.make_clean_working_dir()
+
+        # --no-checkout: Makes it explicit what we are doing here. Nothing is checked out
+        #                until it's explicitly done.
+        # --depth 1: Shallow clone, fetch as little data as possible.
+        cmd = ["git", "clone", "--depth", "1"]
+
+        # We add "--no-checkout" in all git clone operations, except:
+        # There exists a case of version_type=BRANCH without a branch name.
+        # This case is relevant for building projects for the first time without knowing the name
+        # of the default branch.
+        if self.version_type == BRANCH and not self.version_identifier:
+            self._skip_fetch = True
+        else:
+            cmd.append("--no-checkout")
```

Comment:
I would like to keep it since it's a step towards a desired behavior: We do not check out data by accident during the clone step.

The real issue here is that we are doing builds on an unknown branch/tag/commit. In this "exception", we correctly identify the only known "broken" behavior, making it easier to eliminate in the future -- when we have figured out how to reliably trigger the first build after having saved the name of the default branch.

By removing `--no-checkout`, we invite in more of the old undesired behavior instead of forcing a strict pattern.

Reply:
While this is "in test" with the Feature flag switched on for selected projects, I'd also be curious to see what assumptions that might have been missed.

Reply:
I've removed it. I don't like that we aren't trying to be clear about _not_ checking out anything in this step. I've left a comment instead.